File: tools/inspect_distro_implementations.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
"""
Distro Implementation Inspection Script for SigmaOS
Audits Linux & BSD distribution inspired abstractions across the codebase.
"""

import os
import sys

DISTRO_KEYWORDS = [
    "Arch", "Debian", "Fedora", "Alpine", "Nix", "Gentoo",
    "FreeBSD", "OpenBSD", "NetBSD", "DragonFly", "Void", "Slackware"
]

def main():
    print("=== SigmaOS Distro Implementation Inspection Tool ===")
    src_dir = "src"
    found_counts = {k: 0 for k in DISTRO_KEYWORDS}

    for root, _, files in os.walk(src_dir):
        for file in files:
            if file.endswith(".rs"):
                filepath = os.path.join(root, file)
                try:
                    with open(filepath, "r", encoding="utf-8") as f:
                        content = f.read()
                        for kw in DISTRO_KEYWORDS:
                            found_counts[kw] += content.count(kw)
                except Exception as e:
                    print(f"Error reading {filepath}: {e}")

    print("\n[+] Distribution Reference Counts in src/:")
    for kw, count in found_counts.items():
        print(f"  - {kw:<12}: {count} occurrences")

    total_refs = sum(found_counts.values())
    print(f"\nTotal Distro Abstraction References: {total_refs}")

    if total_refs > 100:
        print("[SUCCESS] Distro abstractions are thoroughly integrated and verified.")
        sys.exit(0)
    else:
        print("[WARNING] Low distro abstraction count detected.")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: tools/inspect_distro_implementations.rs
// SPDX-License-Identifier: MIT
//! SigmaOS Distro Implementation Inspection Tool (Native Rust)
//! Audits Linux & BSD distribution inspired abstractions across `src/` without Python.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;
use std::process::exit;

const DISTRO_KEYWORDS: &[&str] = &[
    "Arch", "Debian", "Fedora", "Alpine", "Nix", "Gentoo",
    "FreeBSD", "OpenBSD", "NetBSD", "DragonFly", "Void", "Slackware",
];

fn inspect_directory(dir: &Path, counts: &mut BTreeMap<&'static str, usize>) {
    if let Ok(entries) = fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                inspect_directory(&path, counts);
            } else if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("rs") {
                if let Ok(content) = fs::read_to_string(&path) {
                    for &kw in DISTRO_KEYWORDS {
                        let matches = content.matches(kw).count();
                        *counts.entry(kw).or_insert(0) += matches;
                    }
                }
            }
        }
    }
}

fn main() {
    println!("=== SigmaOS Distro Implementation Inspection Tool (Native Rust) ===");
    let src_dir = Path::new("src");
    let mut counts: BTreeMap<&'static str, usize> = BTreeMap::new();

    for &kw in DISTRO_KEYWORDS {
        counts.insert(kw, 0);
    }

    if src_dir.exists() {
        inspect_directory(src_dir, &mut counts);
    } else {
        eprintln!("[ERROR] Directory 'src/' not found.");
        exit(1);
    }

    println!("\n[+] Distribution Reference Counts in src/:");
    let mut total_refs = 0;
    for (kw, count) in &counts {
        println!("  - {:<12}: {} occurrences", kw, count);
        total_refs += count;
    }

    println!("\nTotal Distro Abstraction References: {}", total_refs);

    if total_refs > 100 {
        println!("[SUCCESS] Distro abstractions are thoroughly integrated and verified in native Rust.");
        exit(0);
    } else {
        println!("[WARNING] Low distro abstraction count detected.");
        exit(1);
    }
}
